Lib/matlab/matlabrun.swg

```diff
@@ -204,65 +270,72 @@ void SWIG_Matlab_ExitFcn(void) {
 
 /* Runtime API implementation */
 SWIGRUNTIME mxArray* SWIG_Matlab_NewPointerObj(void *ptr, swig_type_info *type, int flags) {
-  mxArray* pm_command = mxCreateString("_swigCreate");
-  mxArray* pm;
-  const char * matlab_classname;
-  if (type->clientdata) {
-    matlab_classname = (const char *)type->clientdata;
-  } else {
-    /* TODO do we really want this? it's a fall-back for non-package types (e.g pointer-to-float)
-       However, it will fail anyway because the mexCallMATLABWithTrap call will fail 
-       (as there'd be no package.float matlab-type)
-    */
-    matlab_classname = type->name+3;    /* FIXME: Hack to dereference pointer */
-    mexWarnMsgIdAndTxt("SWIG:NewPointerObj","NewPointerObj problem. clientdata NULL. guessed %s", matlab_classname);
-  }
-
-  mxArray* ex = mexCallMATLABWithTrap(1, &pm, 1, &pm_command, matlab_classname);
-
-  /* Hack hack */
-  if(ex){
-    /* the above failed, so let's just create a SwigRef */
-    mxDestroyArray(ex);
-    mexWarnMsgIdAndTxt("SWIG:NewPointerObj","NewPointerObj problem. creating SwigRef as opposed to %s", matlab_classname);
-    ex = mexCallMATLABWithTrap(1, &pm, 0, 0,"SwigRef");
-  }
-
-  if(ex) mexErrMsgIdAndTxt("SWIG:NewPointerObj","mexCallMATLABWithTrap failed for %s",type->name);
-  mxDestroyArray(pm_command);
+  mxArray *pm = 0;
 
   /* Ownership? */
   int own = (flags & SWIG_POINTER_OWN) ? SWIG_POINTER_OWN : 0;
 
   /* Allocate a pointer object */  
-  size_t ind;
+  size_t ind = 0;
   if (SWIG_Matlab_NewPointer(&ind, ptr, type, own)) {
     mexErrMsgIdAndTxt("SWIG:NewPointerObj","Cannot allocate pointer");
   }
 
-  /* Pass to class */
-  mxArray *pm_ind = mxCreateNumericMatrix(1, 1, mxUINT64_CLASS, mxREAL);
-  *(uint64_T *)mxGetData(pm_ind) = (uint64_T)ind;
-  if(!pm_ind) mexErrMsgIdAndTxt("SWIG:NewPointerObj","mxCreateNumericMatrix failed");
-  mxSetProperty(pm, 0, "swigInd", pm_ind);
-  mxDestroyArray(pm_ind);
+  /* Different behavior depending on ownership */
+  if (!own) {
```

Comment:
Yes, I’m thinking that we should use a global variable to store swigPtr.  The first module to load then manages storage for all other modules.  I can make changes locally in my fork and send a pull request.


> On Aug 29, 2015, at 9:50 AM, Joel Andersson <notifications@github.com> wrote:
> 
> In Lib/matlab/matlabrun.swg <https://github.com/jaeandersson/swig/pull/40#discussion_r38262133>:
> 
> >    if (SWIG_Matlab_NewPointer(&ind, ptr, type, own)) {
> >      mexErrMsgIdAndTxt("SWIG:NewPointerObj","Cannot allocate pointer");
> >    }
> >  
> > -  /* Pass to class */
> > -  mxArray *pm_ind = mxCreateNumericMatrix(1, 1, mxUINT64_CLASS, mxREAL);
> > -  *(uint64_T *)mxGetData(pm_ind) = (uint64_T)ind;
> > -  if(!pm_ind) mexErrMsgIdAndTxt("SWIG:NewPointerObj","mxCreateNumericMatrix failed");
> > -  mxSetProperty(pm, 0, "swigInd", pm_ind);
> > -  mxDestroyArray(pm_ind);
> > +  /* Different behavior depending on ownership */
> > +  if (!own) {
> @dcb210 <https://github.com/dcb210> This design won't work. It is breaking for all function that returns object instances:
> 
> MyWrappedClass myfunction();
> —
> Reply to this email directly or view it on GitHub <https://github.com/jaeandersson/swig/pull/40/files#r38262133>.
>